scripts/export-openapi-cli.js

```diff
@@ -0,0 +1,349 @@
+import fs from 'fs'
+import path from 'path'
+import yaml from 'js-yaml'
+import { collectDefinitions } from '../core/base-service/loader.js'
+
+const baseUrl = process.env.BASE_URL || 'https://img.shields.io'
+const specsPath = path.join('frontend', 'categories')
+
+function getCodeSamples(altText) {
+  return [
+    {
+      lang: 'URL',
+      label: 'URL',
+      source: '$url',
+    },
+    {
+      lang: 'Markdown',
+      label: 'Markdown',
+      source: `![${altText}]($url)`,
+    },
+    {
+      lang: 'reStructuredText',
+      label: 'rSt',
+      source: `.. image:: $url\n:   alt: ${altText}`,
+    },
+    {
+      lang: 'AsciiDoc',
+      label: 'AsciiDoc',
+      source: `image:$url[${altText}]`,
+    },
+    {
+      lang: 'HTML',
+      label: 'HTML',
+      source: `<img alt="${altText}" src="$url">`,
+    },
+  ]
+}
+
+function pattern2openapi(pattern) {
+  return pattern
+    .replace(/:([A-Za-z0-9_\-.]+)(?=[/]?)/g, (matches, grp1) => `{${grp1}}`)
+    .replace(/\([^)]*\)/g, '')
+    .replace(/\+$/, '')
+}
+
+function getEnum(pattern, paramName) {
+  const re = new RegExp(`${paramName}\\(([A-Za-z0-9_\\-|]+)\\)`)
+  const match = pattern.match(re)
+  if (match === null) {
+    return undefined
+  }
+  if (!match[1].includes('|')) {
+    return undefined
+  }
+  return match[1].split('|')
+}
+
+function param2openapi(pattern, paramName, exampleValue, paramType) {
+  const outParam = {}
+  outParam.name = paramName
+  // TODO: outParam.description we don't store this yet
+  outParam.in = paramType
+  if (paramType === 'path') {
+    outParam.required = true
+  } else {
+    /*
+    Occasionally we do have required query params,
+    but at the moment we can't pick this up.
+
+    TODO: Add something to the example format which allows
+    us to populate this based on the queryParamSchema.
+    */
+    outParam.required = false
```

Comment:
There's some stuff we just can't really populate from the existing data structure. Rather than try to extend it, move to allowing services to declare openapi path directly.

Reply:
Could you elaborate on this for me? I'm not confident I understand the logic and comments of the if/else block and the conditionality of `required`

Reply:
I know you resolved this comment, but I think it is worth explaining this one in more detail.

Basically, in our existing `examples` array that we attach to service classes, we don't say if a param is required or optional.
For path (or route) params, this doesn't matter. All path params are required in OpenAPI.
For query params, this does matter because in OpenAPI they can be required or optional, and we do have a few services where there are query params that are required at the schema level.
For the moment I'm just setting them all to optional, which is right in most cases but there will be a handful of cases where we are documenting a required query param as optional in the frontend but then requiring it at the schema level. These will need improving later, but the way I want to do this is by moving to declaring an `openApi` spec object directly on the service classes, rather than extending the existing `examples` object (which already requires some bashing with a hammer to get it into the right format).